src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/RazorCodeLensProvider.ts

```diff
@@ -0,0 +1,130 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import * as vscode from 'vscode';
+import { isRazorCSharpFile, getRazorDocumentUri } from './RazorConventions';
+import { RazorLanguageFeatureBase } from './RazorLanguageFeatureBase';
+import { LanguageKind } from './RPC/LanguageKind';
+import { CancellationToken } from 'vscode-jsonrpc';
+
+class RazorCodeLens extends vscode.CodeLens {
+    constructor(
+        range: vscode.Range,
+        public uri: vscode.Uri,
+        public document: vscode.TextDocument,
+        command?: vscode.Command) {
+
+        super(range, command);
+    }
+}
+
+export class RazorCodeLensProvider
+    extends RazorLanguageFeatureBase
+    implements vscode.CodeLensProvider {
+
+    public async provideCodeLenses(document: vscode.TextDocument, token: vscode.CancellationToken) {
+        try {
+            const razorDocument = await this.documentManager.getDocument(document.uri);
+            if (!razorDocument) {
+                return;
+            }
+
+            const csharpDocument = razorDocument.csharpDocument;
+
+            // Get all the code lenses that applies to our projected C# document.
+            const codeLenses = await vscode.commands.executeCommand<vscode.CodeLens[]>(
```

Comment:
It's possible this returns `undefined` if for instance the underlying codelens action takes too long.

Reply:
Also just to clarify, the reason why we don't need to remap document file paths below is because of the language middleware piece right?

Reply:
> It's possible this returns undefined if for instance the underlying codelens action takes too long.

Interesting. I'll add an if.

> Also just to clarify, the reason why we don't need to remap document file paths below is because of the language middleware piece right?

We do need to remap in this method. We don't need to remap in `resolveCodeLens()` because of language middleware.